tsig_test.go

```diff
@@ -231,9 +232,128 @@ func TestTSIGHMAC224And384(t *testing.T) {
 			if mac != tc.expectedMAC {
 				t.Fatalf("MAC doesn't match: expected '%s' but got '%s'", tc.expectedMAC, mac)
 			}
-			if err = tsigVerify(msgData, tc.secret, "", false, timeSigned); err != nil {
+			if err = tsigVerify(msgData, "", false, timeSigned, tsigHMACProvider(tc.secret)); err != nil {
 				t.Error(err)
 			}
 		})
 	}
 }
+
+const (
+	testGoodKeyName = "goodkey."
+)
+
+var (
+	testErrBadKey = errors.New("this is an intentional error")
+	testGoodMAC   = []byte{0, 1, 2, 3}
+)
+
+// testProvider always generates the same MAC and only accepts the one signature
+type testProvider struct {
+	GenerateAllKeys bool
+}
+
+func (provider *testProvider) Generate(_ []byte, t *TSIG) ([]byte, error) {
+	switch {
+	case t.Hdr.Name == testGoodKeyName || provider.GenerateAllKeys:
+		return testGoodMAC, nil
+	default:
+		return nil, testErrBadKey
+	}
+}
+
+func (*testProvider) Verify(_ []byte, t *TSIG) error {
+	switch t.Hdr.Name {
+	case testGoodKeyName:
+		return nil
+	default:
+		return testErrBadKey
+	}
+}
+
+func TestTsigGenerateProvider(t *testing.T) {
+	tables := map[string]struct {
```

Comment:
Use a slice instead of a map as the randomised map order can be annoying. Also just using `keyname` as the test name would probably be clear enough and fine.

Reply:
Fixed.